File: src/main/java/org/dedda/bratwurst/lang/FunctionCall.java
package org.dedda.bratwurst.lang;

import org.dedda.bratwurst.lang.scope.Scope;

import java.util.Arrays;

/**
 * Created by dedda on 10/16/15.
 *
 * @author dedda
 */
public class FunctionCall extends BWExpression {

    private final String functionName;
    private String variableName = "";
    private BWObject value = new BWInteger(0);

    public FunctionCall(int lineNumber, String functionName, BWVariable[] functionArguments) {
        super(lineNumber);
        this.functionName = functionName;
        setArguments(functionArguments);
    }

    public FunctionCall(int lineNumber, String variableName, String functionName, BWVariable[] functionArguments) {
        super(lineNumber);
        this.variableName = variableName;
        this.functionName = functionName;
        setArguments(functionArguments);
    }

    @Override
    public BWObject getValue() {
        return value;
    }

    @Override
    public int getIntValue() {
        return value.getIntValue();
    }

    @Override
    public String getValueType() {
        return value.getValueType();
    }

    @Override
    public void run(Scope scope) {
        BWFunction function = null;
        if (!variableName.equals("")) {
            function = scope.getFunction(variableName, functionName);
        }
        if (function == null) {
            function = scope.getFunction(functionName);
        }
        if (function == null) {
            throw new RuntimeException(getLineNumber() + ": function " + variableName + "." + functionName + " not set!");
        }
        for (BWVariable argument : getArguments()) {
            argument.run(scope);
        }
        if (!variableName.equals("")) {
//            scope = new Scope(scope.getVariable(variableName).getValue());
            scope.enterFunction(scope.getVariable(variableName).getValue(), function, Arrays.asList(getArguments()));
        } else {
            scope.enterFunction(function, Arrays.asList(getArguments()));
        }
        function.setArguments(getArguments());
        function.run(scope);
        value = function.getValue();
        scope.leaveFunction();
    }
}

File: src/main/java/org/dedda/bratwurst/lang/FunctionCall.kt
package org.dedda.bratwurst.lang

import org.dedda.bratwurst.lang.scope.Scope

/**
 * Created by dedda on 10/16/15.
 *
 * @author dedda
 */
class FunctionCall : BWExpression {
    private val functionName: String
    private var variableName = ""
    private var value: BWObject = BWInteger(0)

    constructor(lineNumber: Int, functionName: String, functionArguments: Array<BWVariable>) : super(lineNumber) {
        this.functionName = functionName
        arguments = functionArguments
    }

    constructor(lineNumber: Int, variableName: String, functionName: String, functionArguments: Array<BWVariable>) : super(lineNumber) {
        this.variableName = variableName
        this.functionName = functionName
        arguments = functionArguments
    }

    override fun getValue(): BWObject {
        return value
    }

    override fun getIntValue(): Int {
        return value.intValue
    }

    override fun getValueType() = value.valueType

    override fun run(scope: Scope) {
        var function: BWFunction? = null
        if (variableName != "") {
            function = scope.getFunction(variableName, functionName)
        }
        if (function == null) {
            function = scope.getFunction(functionName)
        }
        if (function == null) {
            throw RuntimeException("$lineNumber: function $variableName.$functionName not set!")
        }
        for (argument in arguments) {
            argument.run(scope)
        }
        if (variableName != "") {
            scope.enterFunction(scope.getVariable(variableName).value, function, listOf(*arguments))
        } else {
            scope.enterFunction(function, listOf(*arguments))
        }
        function.arguments = arguments
        function.run(scope)
        value = function.value
        scope.leaveFunction()
    }
}